pinot-minion-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/CustomerBasedRetentionTaskGenerator.java

```diff
@@ -0,0 +1,203 @@
+package org.apache.pinot.plugin.minion.tasks;
+
+import com.google.common.base.Preconditions;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.task.TaskState;
+import org.apache.pinot.common.metadata.segment.OfflineSegmentZKMetadata;
+import org.apache.pinot.controller.helix.core.minion.ClusterInfoAccessor;
+import org.apache.pinot.controller.helix.core.minion.generator.PinotTaskGenerator;
+import org.apache.pinot.controller.helix.core.minion.generator.TaskGeneratorUtils;
+import org.apache.pinot.core.common.MinionConstants;
+import org.apache.pinot.core.minion.PinotTaskConfig;
+import org.apache.pinot.spi.annotations.minion.TaskGenerator;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.TableTaskConfig;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.utils.TimeUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@TaskGenerator
+public class CustomerBasedRetentionTaskGenerator implements PinotTaskGenerator{
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(CustomerBasedRetentionTaskGenerator.class);
+  private static final String TASK_TYPE = "customerBasedRetentionTask";
+  public static final String BUCKET_TIME_PERIOD_KEY = "bucketTimePeriod";
+  private static final String CUSTOMER_RETENTION_CONFIG = "customerRetentionConfig";
+  public static final String WINDOW_START_MS_KEY = "windowStartMs";
+  public static final String WINDOW_END_MS_KEY = "windowEndMs";
+
+  private ClusterInfoAccessor _clusterInfoAccessor;
+
+  @Override
+  public void init(ClusterInfoAccessor clusterInfoAccessor) {
+    _clusterInfoAccessor = clusterInfoAccessor;
+  }
+
+  @Override
+  public String getTaskType() {
+    return TASK_TYPE;
+  }
+
+  @Override
+  public List<PinotTaskConfig> generateTasks(List<TableConfig> tableConfigs) {
+    List<PinotTaskConfig> pinotTaskConfigs = new ArrayList<>();
+
+    for (TableConfig tableConfig : tableConfigs) {
+      String offlineTableName = tableConfig.getTableName();
+
+      if (tableConfig.getTableType() != TableType.OFFLINE) {
+        LOGGER.warn("Skip generating task: {} for non-OFFLINE table: {}", TASK_TYPE, offlineTableName);
+        continue;
+      }
+
+      LOGGER.info("Start generating task configs for table: {} for task: {}", offlineTableName, TASK_TYPE);
+
+      // Only schedule 1 job of this type (this optimisation may be moved downstream later on)
+      Map<String, TaskState> incompleteTasks =
```

Comment:
@laxmanchekka instead of this, can we use running segment check similar to [this](https://github.com/hypertrace/pinot/pull/98#discussion_r782188846) ?